notebooks/COS/ExceptionReport/cos_functions.py

```diff
@@ -0,0 +1,434 @@
+    for i in range(len(data_table)):
+        wvln_, gross_, gcount_ = data_table[i]["WAVELENGTH", "GROSS", "GCOUNTS"]
+
+        # No range specified - estimates over the whole range
+        snr_array.append([wvln_, np.sqrt(gcount_), i])
+        if snr_range == [-1, -1]:
+            if verbose:
+                print("No range specified.")
+
+        else:
+            if (min(snr_range) > min(wvln_)) & (max(snr_range) < max(wvln_)):
+                segsFound += 1
+                wvln_range_mask = (wvln_ > snr_range[0]) & (wvln_ < snr_range[1])
+
+                wvln_range, gcount_range, gross_range = (
+                    wvln_[wvln_range_mask],
+                    gcount_[wvln_range_mask],
+                    gross_[wvln_range_mask],
+                )
+
+                snr_array.append([wvln_range, np.sqrt(gcount_range), i])
+
+                if weighted is False:
+                    weight_avg_snr = np.mean(np.sqrt(gcount_range))
+                    if verbose:
+                        print(
+                            f"In range on {i}-th segment with limits:",
+                            min(wvln_),
+                            max(wvln_),
+                            f"\nUnweighted mean SNR over the range {snr_range} is: {weight_avg_snr}",
+                        )
+
+                if weighted:
+                    weight_avg_snr = np.average(
+                        np.sqrt(gcount_range),
+                        weights=np.nan_to_num(gcount_range / gross_range, nan=1e-30),
+                    )
+                    if verbose:
+                        print(
+                            f"In range on {i}-th segment with limits:",
+                            min(wvln_),
+                            max(wvln_),
+                            f"\nEXPTIME weighted average SNR over the range {snr_range} is: {weight_avg_snr}",
+                        )
+            else:
+                snr_array.append([-1, -1, -1])
+                if verbose:
+                    print(
+                        f"Out of range on {i}-th segment with limits:",
+                        min(wvln_),
+                        max(wvln_),
+                    )
+    if (  # Check whether the code has found any specified wavelength ranges
+        all([elem == -1 for elem in (snr_array[0][2], snr_array[1][2])])
+    ) & (  # Check that the user specified wavelength ranges
+        snr_range != [-1, -1]
+    ):
+        if verbose:  # Then print to the user that the ranges weren't found
+            print("\nThe input range was not found in any segment!\n")
+    if segsFound > 1:
+        if verbose:
+            print(
+                "\nThis range was found on multiple segments, (grating = G230L?) ,which at present is not fully supported. The returned array should be accurate, but the mean may be incorrect."
+            )
+    return weight_avg_snr, snr_array
+
+
+# %%
+def withinPercent(val1, val2, percent=1.0):
+    """
```

Comment:
I don't see this being called anywhere. Could you please add instructions on how to test in the notebook or consider removing it entirely?